shared/messagestore/src/main/kotlin/company/thewebhook/messagestore/producer/Producer.kt

```diff
@@ -12,3 +14,10 @@ abstract class Producer<T> {
     abstract suspend fun publish(topic: String, message: T): Boolean
     abstract suspend fun close()
 }
+
+abstract class DelayedProducer<T> : Producer<T>() {
+    abstract suspend fun publish(topic: String, message: T, delay: Duration): Boolean
+    override suspend fun publish(topic: String, message: T): Boolean {
+        return publish(topic, message, 0.seconds)
+    }
+}
```

Comment:
Won't an interface be a better approach? What if we had a different variety of producer as well, and we wanted to mix both varieties?

Delayed messages is a capability of a Producer. So if we had an interface, say "DelayedMessages" (can think of a better name). We can always assert for this capability using generics:
(in Java, I'm sure Kotlin would prolly have something similar)
```
T extends Producer implements DelayedMessages
```

The advantage is, if we had some other capabilities which we need, we can add it on to the list of implementations, like:
```
T extends Producer implements DelayedMessages, SomeOtherCapability
```

Reply:
yep makes sense